File: python/steering_functions/dubins.py
from __future__ import annotations

from dataclasses import dataclass
import math
from typing import List, Sequence, Tuple

from .core import BaseStateSpace, Control, State, twopify

DUBINS_EPS = 1e-6
DUBINS_ZERO = -1e-9

LEFT = "L"
STRAIGHT = "S"
RIGHT = "R"

_DUBINS_PATH_TYPES: tuple[tuple[str, str, str], ...] = (
    (LEFT, STRAIGHT, LEFT),
    (RIGHT, STRAIGHT, RIGHT),
    (RIGHT, STRAIGHT, LEFT),
    (LEFT, STRAIGHT, RIGHT),
    (RIGHT, LEFT, RIGHT),
    (LEFT, RIGHT, LEFT),
)


@dataclass
class DubinsPath:
    type_: Sequence[str] = _DUBINS_PATH_TYPES[0]
    lengths: Tuple[float, float, float] = (0.0, math.inf, 0.0)

    def length(self) -> float:
        return sum(self.lengths)


def _dubins_lsl(d: float, alpha: float, beta: float) -> DubinsPath:
    ca, sa, cb, sb = math.cos(alpha), math.sin(alpha), math.cos(beta), math.sin(beta)
    tmp = 2.0 + d * d - 2.0 * (ca * cb + sa * sb - d * (sa - sb))
    if tmp >= DUBINS_ZERO:
        theta = math.atan2(cb - ca, d + sa - sb)
        t = twopify(-alpha + theta)
        p = math.sqrt(max(tmp, 0.0))
        q = twopify(beta - theta)
        return DubinsPath(_DUBINS_PATH_TYPES[0], (t, p, q))
    return DubinsPath()


def _dubins_rsr(d: float, alpha: float, beta: float) -> DubinsPath:
    ca, sa, cb, sb = math.cos(alpha), math.sin(alpha), math.cos(beta), math.sin(beta)
    tmp = 2.0 + d * d - 2.0 * (ca * cb + sa * sb - d * (sb - sa))
    if tmp >= DUBINS_ZERO:
        theta = math.atan2(ca - cb, d - sa + sb)
        t = twopify(alpha - theta)
        p = math.sqrt(max(tmp, 0.0))
        q = twopify(-beta + theta)
        return DubinsPath(_DUBINS_PATH_TYPES[1], (t, p, q))
    return DubinsPath()


def _dubins_rsl(d: float, alpha: float, beta: float) -> DubinsPath:
    ca, sa, cb, sb = math.cos(alpha), math.sin(alpha), math.cos(beta), math.sin(beta)
    tmp = d * d - 2.0 + 2.0 * (ca * cb + sa * sb - d * (sa + sb))
    if tmp >= DUBINS_ZERO:
        p = math.sqrt(max(tmp, 0.0))
        theta = math.atan2(ca + cb, d - sa - sb) - math.atan2(2.0, p)
        t = twopify(alpha - theta)
        q = twopify(beta - theta)
        return DubinsPath(_DUBINS_PATH_TYPES[2], (t, p, q))
    return DubinsPath()


def _dubins_lsr(d: float, alpha: float, beta: float) -> DubinsPath:
    ca, sa, cb, sb = math.cos(alpha), math.sin(alpha), math.cos(beta), math.sin(beta)
    tmp = -2.0 + d * d + 2.0 * (ca * cb + sa * sb + d * (sa + sb))
    if tmp >= DUBINS_ZERO:
        p = math.sqrt(max(tmp, 0.0))
        theta = math.atan2(-ca - cb, d + sa + sb) - math.atan2(-2.0, p)
        t = twopify(-alpha + theta)
        q = twopify(-beta + theta)
        return DubinsPath(_DUBINS_PATH_TYPES[3], (t, p, q))
    return DubinsPath()


def _dubins_rlr(d: float, alpha: float, beta: float) -> DubinsPath:
    ca, sa, cb, sb = math.cos(alpha), math.sin(alpha), math.cos(beta), math.sin(beta)
    tmp = 0.125 * (6.0 - d * d + 2.0 * (ca * cb + sa * sb + d * (sa - sb)))
    if abs(tmp) < 1.0:
        p = 2.0 * math.pi - math.acos(tmp)
        theta = math.atan2(ca - cb, d - sa + sb)
        t = twopify(alpha - theta + 0.5 * p)
        q = twopify(alpha - beta - t + p)
        return DubinsPath(_DUBINS_PATH_TYPES[4], (t, p, q))
    return DubinsPath()


def _dubins_lrl(d: float, alpha: float, beta: float) -> DubinsPath:
    ca, sa, cb, sb = math.cos(alpha), math.sin(alpha), math.cos(beta), math.sin(beta)
    tmp = 0.125 * (6.0 - d * d + 2.0 * (ca * cb + sa * sb - d * (sa - sb)))
    if abs(tmp) < 1.0:
        p = 2.0 * math.pi - math.acos(tmp)
        theta = math.atan2(-ca + cb, d + sa - sb)
        t = twopify(-alpha + theta + 0.5 * p)
        q = twopify(beta - alpha - t + p)
        return DubinsPath(_DUBINS_PATH_TYPES[5], (t, p, q))
    return DubinsPath()


def _select_dubins_path(d: float, alpha: float, beta: float) -> DubinsPath:
    if d < DUBINS_EPS and abs(alpha - beta) < DUBINS_EPS:
        return DubinsPath(_DUBINS_PATH_TYPES[0], (0.0, d, 0.0))

    candidates = [
        _dubins_lsl(d, alpha, beta),
        _dubins_rsr(d, alpha, beta),
        _dubins_rsl(d, alpha, beta),
        _dubins_lsr(d, alpha, beta),
        _dubins_rlr(d, alpha, beta),
        _dubins_lrl(d, alpha, beta),
    ]
    return min(candidates, key=lambda path: path.length())


class DubinsStateSpace(BaseStateSpace):
    def __init__(self, kappa: float, discretization: float = 0.1, forwards: bool = True) -> None:
        super().__init__(discretization)
        if kappa <= 0.0:
            raise ValueError("kappa must be positive")
        self._kappa = kappa
        self._kappa_inv = 1.0 / kappa
        self._forwards = forwards

    def dubins(self, state1: State, state2: State) -> DubinsPath:
        dx = state2.x - state1.x
        dy = state2.y - state1.y
        theta = math.atan2(dy, dx)
        d = math.hypot(dx, dy) * self._kappa
        alpha = twopify(state1.theta - theta)
        beta = twopify(state2.theta - theta)
        return _select_dubins_path(d, alpha, beta)

    def get_distance(self, state1: State, state2: State) -> float:
        if self._forwards:
            return self._kappa_inv * self.dubins(state1, state2).length()
        return self._kappa_inv * self.dubins(state2, state1).length()

    def _get_controls_core(
        self,
        state1: State,
        state2: State,
        reverse_direction: bool = False,
    ) -> List[Control]:
        path = self.dubins(state2, state1) if reverse_direction else self.dubins(state1, state2)
        controls: List[Control] = []

        for segment_type, length in zip(path.type_, path.lengths):
            if segment_type == LEFT:
                control = Control(self._kappa_inv * length, self._kappa, 0.0)
            elif segment_type == RIGHT:
                control = Control(self._kappa_inv * length, -self._kappa, 0.0)
            else:
                control = Control(self._kappa_inv * length, 0.0, 0.0)

            if abs(control.delta_s) > DUBINS_EPS:
                controls.append(control)

        if reverse_direction:
            controls.reverse()
            controls = [Control(-control.delta_s, control.kappa, control.sigma) for control in controls]

        return controls

    def get_controls(self, state1: State, state2: State) -> List[Control]:
        return self._get_controls_core(state1, state2, False)

    def get_controls_reverse(self, state1: State, state2: State) -> List[Control]:
        return self._get_controls_core(state1, state2, True)

    def get_all_controls(self, state1: State, state2: State) -> List[List[Control]]:
        return [self.get_controls(state1, state2), self.get_controls_reverse(state1, state2)]

    def get_path(self, state1: State, state2: State) -> List[State]:
        forward_length = self.get_distance(state1, state2)
        reverse_length = self._kappa_inv * self.dubins(state2, state1).length()
        controls = (
            self._get_controls_core(state1, state2, False)
            if forward_length <= reverse_length
            else self._get_controls_core(state1, state2, True)
        )
        return self.integrate(state1, controls)

File: rust/src/dubins.rs
use crate::state::{State, Control};
use crate::base_state_space::StateSpace;
use crate::utilities::twopify;

const DUBINS_LEFT: u8 = 0;
const DUBINS_STRAIGHT: u8 = 1;
const DUBINS_RIGHT: u8 = 2;

const DUBINS_PATH_TYPE: [[u8; 3]; 6] = [
    [DUBINS_LEFT, DUBINS_STRAIGHT, DUBINS_LEFT],
    [DUBINS_RIGHT, DUBINS_STRAIGHT, DUBINS_RIGHT],
    [DUBINS_RIGHT, DUBINS_STRAIGHT, DUBINS_LEFT],
    [DUBINS_LEFT, DUBINS_STRAIGHT, DUBINS_RIGHT],
    [DUBINS_RIGHT, DUBINS_LEFT, DUBINS_RIGHT],
    [DUBINS_LEFT, DUBINS_RIGHT, DUBINS_LEFT],
];

const DUBINS_EPS: f64 = 1e-6;
const DUBINS_ZERO: f64 = -1e-9;

#[derive(Clone, Debug)]
struct DubinsPath {
    type_: [u8; 3],
    length_: [f64; 3],
}

impl DubinsPath {
    fn new(type_: [u8; 3], length_: [f64; 3]) -> Self {
        Self { type_, length_ }
    }
    fn invalid() -> Self {
        Self {
            type_: DUBINS_PATH_TYPE[0],
            length_: [0.0, f64::INFINITY, 0.0],
        }
    }
    fn length(&self) -> f64 {
        self.length_[0] + self.length_[1] + self.length_[2]
    }
}

fn dubins_lsl(d: f64, alpha: f64, beta: f64) -> DubinsPath {
    let (ca, sa) = (alpha.cos(), alpha.sin());
    let (cb, sb) = (beta.cos(), beta.sin());
    let tmp = 2.0 + d * d - 2.0 * (ca * cb + sa * sb - d * (sa - sb));
    if tmp >= DUBINS_ZERO {
        let theta = (cb - ca).atan2(d + sa - sb);
        let t = twopify(-alpha + theta);
        let p = tmp.max(0.0).sqrt();
        let q = twopify(beta - theta);
        DubinsPath::new(DUBINS_PATH_TYPE[0], [t, p, q])
    } else {
        DubinsPath::invalid()
    }
}

fn dubins_rsr(d: f64, alpha: f64, beta: f64) -> DubinsPath {
    let (ca, sa) = (alpha.cos(), alpha.sin());
    let (cb, sb) = (beta.cos(), beta.sin());
    let tmp = 2.0 + d * d - 2.0 * (ca * cb + sa * sb - d * (sb - sa));
    if tmp >= DUBINS_ZERO {
        let theta = (ca - cb).atan2(d - sa + sb);
        let t = twopify(alpha - theta);
        let p = tmp.max(0.0).sqrt();
        let q = twopify(-beta + theta);
        DubinsPath::new(DUBINS_PATH_TYPE[1], [t, p, q])
    } else {
        DubinsPath::invalid()
    }
}

fn dubins_rsl(d: f64, alpha: f64, beta: f64) -> DubinsPath {
    let (ca, sa) = (alpha.cos(), alpha.sin());
    let (cb, sb) = (beta.cos(), beta.sin());
    let tmp = d * d - 2.0 + 2.0 * (ca * cb + sa * sb - d * (sa + sb));
    if tmp >= DUBINS_ZERO {
        let p = tmp.max(0.0).sqrt();
        let theta = (ca + cb).atan2(d - sa - sb) - (2.0_f64).atan2(p);
        let t = twopify(alpha - theta);
        let q = twopify(beta - theta);
        DubinsPath::new(DUBINS_PATH_TYPE[2], [t, p, q])
    } else {
        DubinsPath::invalid()
    }
}

fn dubins_lsr(d: f64, alpha: f64, beta: f64) -> DubinsPath {
    let (ca, sa) = (alpha.cos(), alpha.sin());
    let (cb, sb) = (beta.cos(), beta.sin());
    let tmp = -2.0 + d * d + 2.0 * (ca * cb + sa * sb + d * (sa + sb));
    if tmp >= DUBINS_ZERO {
        let p = tmp.max(0.0).sqrt();
        let theta = (-ca - cb).atan2(d + sa + sb) - (-2.0_f64).atan2(p);
        let t = twopify(-alpha + theta);
        let q = twopify(-beta + theta);
        DubinsPath::new(DUBINS_PATH_TYPE[3], [t, p, q])
    } else {
        DubinsPath::invalid()
    }
}

fn dubins_rlr(d: f64, alpha: f64, beta: f64) -> DubinsPath {
    let (ca, sa) = (alpha.cos(), alpha.sin());
    let (cb, sb) = (beta.cos(), beta.sin());
    let tmp = 0.125 * (6.0 - d * d + 2.0 * (ca * cb + sa * sb + d * (sa - sb)));
    if tmp.abs() <= 1.0 {
        let p = twopify(std::f64::consts::TAU - tmp.acos());
        let theta = (ca - cb).atan2(d - sa + sb);
        let t = twopify(alpha - theta + 0.5 * p);
        let q = twopify(alpha - beta - t + p);
        DubinsPath::new(DUBINS_PATH_TYPE[4], [t, p, q])
    } else {
        DubinsPath::invalid()
    }
}

fn dubins_lrl(d: f64, alpha: f64, beta: f64) -> DubinsPath {
    let (ca, sa) = (alpha.cos(), alpha.sin());
    let (cb, sb) = (beta.cos(), beta.sin());
    let tmp = 0.125 * (6.0 - d * d + 2.0 * (ca * cb + sa * sb - d * (sa - sb)));
    if tmp.abs() <= 1.0 {
        let p = twopify(std::f64::consts::TAU - tmp.acos());
        let theta = (-ca + cb).atan2(d + sa - sb);
        let t = twopify(-alpha + theta + 0.5 * p);
        let q = twopify(beta - alpha - t + p);
        DubinsPath::new(DUBINS_PATH_TYPE[5], [t, p, q])
    } else {
        DubinsPath::invalid()
    }
}

fn dubins_word(path_type: usize, d: f64, alpha: f64, beta: f64) -> DubinsPath {
    match path_type {
        0 => dubins_lsl(d, alpha, beta),
        1 => dubins_rsr(d, alpha, beta),
        2 => dubins_rsl(d, alpha, beta),
        3 => dubins_lsr(d, alpha, beta),
        4 => dubins_rlr(d, alpha, beta),
        5 => dubins_lrl(d, alpha, beta),
        _ => DubinsPath::invalid(),
    }
}

fn shortest_dubins_path(
    q0: &State, q1: &State, rho: f64, forward: bool,
) -> (DubinsPath, f64) {
    let dx = q1.x - q0.x;
    let dy = q1.y - q0.y;
    let d = dx.hypot(dy) / rho;
    let (theta0, theta1) = if forward {
        (q0.theta, q1.theta)
    } else {
        (q0.theta + std::f64::consts::PI, q1.theta + std::f64::consts::PI)
    };
    let alpha = twopify(theta0 - dy.atan2(dx));
    let beta  = twopify(theta1 - dy.atan2(dx));

    let mut best = DubinsPath::invalid();
    let mut best_len = f64::INFINITY;
    for i in 0..6 {
        let p = dubins_word(i, d, alpha, beta);
        let l = p.length();
        if l < best_len {
            best_len = l;
            best = p;
        }
    }
    (best, best_len * rho)
}

fn controls_from_dubins(path: &DubinsPath, rho: f64, forward: bool, dx: f64, dy: f64) -> Vec<Control> {
    let d = if forward { 1.0 } else { -1.0 };
    let mut controls = Vec::new();
    for i in 0..3 {
        let seg_type = path.type_[i];
        let length = path.length_[i] * rho;
        if length.abs() < DUBINS_EPS { continue; }
        let (kappa, sigma) = match seg_type {
            t if t == DUBINS_LEFT     => (1.0 / rho,  0.0),
            t if t == DUBINS_STRAIGHT => (0.0,         0.0),
            _                         => (-1.0 / rho, 0.0),
        };
        controls.push(Control { delta_s: d * length, kappa, sigma });
    }
    controls
}

/// Dubins state space.
pub struct DubinsStateSpace {
    kappa_: f64,
    discretization_: f64,
    forward_: bool,
}

impl DubinsStateSpace {
    pub fn new(kappa: f64, discretization: f64, forward: bool) -> Self {
        assert!(kappa > 0.0 && discretization > 0.0);
        Self { kappa_: kappa, discretization_: discretization, forward_: forward }
    }
}

impl StateSpace for DubinsStateSpace {
    fn get_controls(&self, s1: &State, s2: &State) -> Vec<Control> {
        let rho = 1.0 / self.kappa_;
        let dx = s2.x - s1.x;
        let dy = s2.y - s1.y;
        let (path, _) = shortest_dubins_path(s1, s2, rho, self.forward_);
        controls_from_dubins(&path, rho, self.forward_, dx, dy)
    }

    fn get_all_controls(&self, s1: &State, s2: &State) -> Vec<Vec<Control>> {
        let rho = 1.0 / self.kappa_;
        let dx = s2.x - s1.x;
        let dy = s2.y - s1.y;
        let mut all = Vec::new();
        let d = dx.hypot(dy) / rho;
        let alpha = twopify(s1.theta - dy.atan2(dx));
        let beta  = twopify(s2.theta - dy.atan2(dx));
        for i in 0..6 {
            let p = dubins_word(i, d, alpha, beta);
            if p.length().is_finite() {
                all.push(controls_from_dubins(&p, rho, self.forward_, dx, dy));
            }
        }
        all
    }

    fn discretization(&self) -> f64 { self.discretization_ }
}
